lib/new_relic/agent/instrumentation/redis/instrumentation.rb

```diff
@@ -6,38 +6,58 @@ module NewRelic::Agent::Instrumentation
   module Redis
     PRODUCT_NAME = 'Redis'
     CONNECT = 'connect'
-    UNKNOWN = "unknown"
-    LOCALHOST = "localhost"
+    UNKNOWN = 'unknown'
+    LOCALHOST = 'localhost'
     MULTI_OPERATION = 'multi'
     PIPELINE_OPERATION = 'pipeline'
+    REDIS_5 = Gem::Version.new(::Redis::VERSION) >= Gem::Version.new('5.0.0') && !!defined?(::RedisClient)
 
+    def connect_with_tracing
+      with_tracing(CONNECT, database: db) { yield }
+    end
+
+    # Used for Redis 4.x and 3.x
     def call_with_tracing(command, &block)
       operation = command[0]
       statement = ::NewRelic::Agent::Datastores::Redis.format_command(command)
 
-      with_tracing(operation, statement) { yield }
+      with_tracing(operation, statement: statement, database: db) { yield }
     end
 
+    # Used for Redis 4.x and 3.x
     def call_pipeline_with_tracing(pipeline)
       operation = pipeline.is_a?(::Redis::Pipeline::Multi) ? MULTI_OPERATION : PIPELINE_OPERATION
       statement = ::NewRelic::Agent::Datastores::Redis.format_pipeline_commands(pipeline.commands)
 
-      with_tracing(operation, statement) { yield }
+      with_tracing(operation, statement: statement, database: db) { yield }
     end
 
-    def connect_with_tracing
-      with_tracing(CONNECT) { yield }
+    # Used for Redis 5.x
+    def call_pipelined_with_tracing(pipeline)
+      operation = pipeline.flatten.include?('MULTI') ? MULTI_OPERATION : PIPELINE_OPERATION
+      statement = ::NewRelic::Agent::Datastores::Redis.format_pipeline_commands(pipeline)
+
+      # call_pipelined isn't invoked on the client object, so use client.db to
+      # access the client instance var on self
+      with_tracing(operation, statement: statement, database: client.db) { yield }
+    end
+
+    # Used for Redis 5.x
+    def call_v_with_tracing(command, &block)
```

Comment:
It looks like the methods `call_v_with_tracing` and `call_with_tracing` are identical in everything other than method name. If i'm understanding correctly, maybe we don't need to have both methods. perhaps we could get rid of this method and just use call_with_tracing for both. so chain could be something like 
```ruby
if method_defined?(:call_v)
  alias_method(:call_v_without_new_relic, :call_v)

  def call_v(*args, &block)
    call_with_tracing(args[0]) { call_v_without_new_relic(*args, &block) }
  end
end

if method_defined?(:call)
  alias_method(:call_without_new_relic, :call)

  def call(*args, &block)
    call_with_tracing(args[0]) { call_without_new_relic(*args, &block) }
  end
end
```

Reply:
You're absolutely right! They were better off different when the middleware was used because we needed to access the database and command differently, but using prepend, they're the same.

Fixed in: 5a66d239247527bc75ca7a5338fb46968387e29e